src/nxt_otel.c

```diff
@@ -0,0 +1,273 @@
+
+/*
+ * Copyright (C) F5, Inc.
+ */
+
+#include <math.h>
+
+#include <nxt_router.h>
+#include <nxt_http.h>
+#include <nxt_otel.h>
+#include <nxt_mp.h>
+#include <nxt_work_queue.h>
+#include <nxt_main.h>
+
+#define NXT_OTEL_TRACEPARENT_LEN 55
+#define NXT_OTEL_BODY_SIZE_TAG "body size"
+
+static inline void nxt_otel_trace_and_span_init(nxt_task_t *, nxt_http_request_t *);
+static inline void nxt_otel_span_collect(nxt_task_t *, nxt_http_request_t *);
+static void nxt_otel_span_add_headers(nxt_task_t *, nxt_http_request_t *);
+static void nxt_otel_span_add_body(nxt_http_request_t *);
+static void nxt_otel_error(nxt_task_t *, nxt_http_request_t *);
+
+inline void
+nxt_otel_test_and_call_state(nxt_task_t *t, nxt_http_request_t *r)
+{
```

Comment:
So, extern declaration here, and inline definition in the header?

Reply:
Yeah... let me see if can dig out a reference for you...

No 3 (A C99 model. Use inline in a common header, ..) towards the end of this [page](https://www.greenend.org.uk/rjk/tech/inline.html).

Reply:
Unfortunately I cannot do that without creating a circular include issue between nxt_otel.h and nxt_http.h.

If it is that important to put the inline definitions into the header I will simply un-inline the functions so that we can leave them in the .c file.

please let me know what you prefer.

Reply:
Hmm, I'd have thought the issue would have been that you're calling static functions that won't be visible outside that TU (other than initial issue of the compiler not even knowing what the functions are without providing prototypes).

I'd say un-inline it for now... it can be revisited later...

The functions being called are either not inline themselves or they are but they are calling other functions which aren't and any possible performace gain is probably lost in the noise...